text/0000-destructor-guarantees.md

```diff
@@ -0,0 +1,176 @@
+> addition to `thread::scoped`.
+
+[scoped-bug]: https://github.com/rust-lang/rust/issues/24292
+[forget-bug]: https://github.com/rust-lang/rust/issues/24456
+[dtor-comment]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93505374
+[dtor-bug1]: https://github.com/rust-lang/rust/issues/14875
+[dtor-bug2]: https://github.com/rust-lang/rust/issues/16135
+[drain-bug]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93513451
+
+Previously, Rust provided no guarantee that any destructors for any types will
+run. However, some of the current APIs (namely `thread::scoped`) were designed
+without keeping this in mind, and can be made memory unsafe through leaking.
+
+This RFC proposes a small, orthogonal feature to allow wondrous RAII-based APIs
+like `thread::scoped` to exist in safe Rust, with minimal breaking changes in
+the pre-1.0 release.
+
+## Narrowing Goals and Dispelling Fear, Uncertainty, and Doubt
+
+As this issue touched a fairly popular and touted API (`thread::scoped`) there
+has naturally been a gigantic amount of discussion and ideation surrounding it.
+
+Some of the proposed solutions to this problem have been far too restrictive,
+such as banning `Rc` cycles entirely, making `Rc` unsafe, or having `Rc`
+require `'static`. This had lead to a certain amount of FUD about the validity
+of all non-safe-`mem::forget` solutions. This RFC is here to dispel this illusion.
+
+### This Proposal Will Not:
+
+ - Prevent *all* destructors from not running.
+ - Prevent destructors from not running in the face of process aborts.
+ - Prevent destructors from not running for `'static` data.
+ - Prevent `Rc` or `Arc` cycles, or ban non-`'static` data from them.
+
+### This Proposal Will:
+
+ - Allow types to manually opt-in to not being leaked.
+ - Allow leaks of `'static` data in a large variety of ways, even if they do
+   not implement `Leak`.
+ - Prevent types which have opted-in to not being leaked from being used in
+   contexts where they might leak.
+ - Prevent panics in destructors from causing leaks or other memory unsafety.
+ - Allow `Rc` and `Arc` cycles, including when they contain non-`'static` data.
+
+# Detailed design
+
+Introduce a new default trait to `core`, `Leak`:
+
+```rust
+pub unsafe trait Leak { }
+unsafe impl Leak for .. { }
+```
+
+Change several `std` APIs to adjust for the guarantees now provided to types
+which do not implement `Leak`:
+
+ - `impl<'a, T> !Leak for ::std::thread::JoinGuard<'a, T> { }`
+   - If we gain some form of specialization, implement `Leak` for `JoinGuard<'static, T>`
+ - implement `!Leak` for the guards returned by collection `drain` operations
+ - Add a `Leak` bound to the type parameter of:
+   - `rc::Rc`
+   - `sync::Arc`
+   - `sync::mpsc::Sender` (possibly not, see unresolved questions)
+   - Possibly other APIs. Please point any others out if you think of them.
+
+_Cause all panics in destructors to immediately abort the process._
```

Comment:
@nagisa This would not work, because:

(1) To be able to rely on this for memory safety it must be a language-wide guarantee.
(2) Destructors may call functions that are not destructors.

Panics within destructors can *already* cause your program to abort, and in fact *will* do so if they occur after another panic (Rust's exception implementation does not allow for safe handling of double panics).  You cannot rely on them not to do this in current Rust.  Furthermore, there are plans to allow whole-program toggling to entirely disable landing pads.  As a result, panics in destructors are already a bad idea and I strongly advise you not to rely on them.